Library/Homebrew/upgrade.rb

```diff
@@ -167,7 +172,7 @@ def upgrade_formula(formula_installer, dry_run: false, verbose: false)
         return
       end
 
-      formula_installer.prelude
+      formula_installer.check_installation_already_attempted
```

Comment:
Can you explain a bit more about this change?

Reply:
Since `prelude` already gets invoked before `fetch`, here we just need to check for `FormulaInstallationAlreadyAttemptedError`. In a previous draft, I made this a boolean method since the error was always being rescued—except in `Cask::Installer` (which is probably a bug).

Reply:
> here we just need to check for `FormulaInstallationAlreadyAttemptedError`

Thanks. Sorry, why do we need to check this?

Reply:
It seems that that error gets raised when we've already attempted to upgrade the formula as part of the dependency tree of another formula. #12018 most likely makes this unnecessary for `brew upgrade`, but there's still `brew install`, `brew reinstall`, and `Cask::Installer` that need to be updated.